metasrv/src/api/http/v1/cluster_state.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use common_exception::Result;
+use poem::http::StatusCode;
+use poem::web::Data;
+use poem::web::Html;
+use poem::web::IntoResponse;
+use poem::Response;
+
+use crate::meta_service::MetaNode;
+
+pub struct ClusterTemplate {
+    result: Result<String>,
+}
+
+impl IntoResponse for ClusterTemplate {
+    fn into_response(self) -> Response {
+        match self.result {
+            Ok(nodes) => Html(nodes).into_response(),
+            Err(cause) => Response::builder()
+                .status(StatusCode::INTERNAL_SERVER_ERROR)
+                .body(format!(
+                    "Failed to fetch cluster nodes list. cause: {}",
+                    cause
+                )),
+        }
+    }
+}
+
+// GET /v1/cluster/nodes
+// list all nodes in current databend-metasrv cluster
+// request: None
+// return: return a list of cluster node information
+#[poem::handler]
+pub async fn nodes_handler(meta_node: Data<&Arc<MetaNode>>) -> ClusterTemplate {
```

Comment:
maybe it can return an Result<String, PoemError> ?

the error processing might be easier with an `Result`

Reply:
@flaneur2020 agree!

Reply:
the rust http library implemented an `IntoResponse` for `Result<T, E>` when `T` and `E` are both implemented `IntoResponse`, like below:

```
impl IntoResponse for Result<T, E> where T: IntoResposne, E: IntoResponse
```

it's already builtin both in axum and poem, do not need us to `impl` this by hand:

https://docs.rs/poem/0.8.0/poem/web/trait.IntoResponse.html#impl-IntoResponse-for-Result%3CT%2C%20E%3E
https://docs.rs/axum/0.1.3/axum/response/trait.IntoResponse.html#impl-IntoResponse-for-Result%3CT%2C%20E%3E